### PR TITLE
network: TCP proxying over HTTP POST

### LINKED ISSUE #14234: Support TCP Proxy over HTTP/2 POST (Encap and Decap)
*Title*: *Support TCP Proxy over HTTP/2 POST*

*Description*:
https://github.com/envoyproxy/envoy/issues/13185 has original context. With https://github.com/envoyproxy/envoy/pull/13548, Envoy could supports the encapping TCP over HTTP/2 POST with a custom TCP upstream.

Based on our further investigation, we notice that in addition to a custom TCP upstream  for encapping for the client side envoy, we also need to a symmetric decapping filter on the server side envoy.

We would like to contribute to Envoy upstream because we think it will benefit other Envoy users as well, as many edge proxies don't have good support for HTTP/2 CONNECT (for example, Google edge frontend). 

If this proposal is accepted, I would like some guidance of 

1. whether the custom upstream should be an independent class or it can be an option to the existing tcp upstream?

2. which directory is the proper place for adding the custom upstream (if needed) and the new filter?

### PR DESCRIPTION
**Pull request by wlhee**
Commit Message: Add support proxying TCP over HTTP/2 or HTTP/1.1 POST to tcp_proxy. It's controlled by tcp_proxy filter configuration. Also add support adding additional headers to the proxy requests to coordinate with the upstream. In addition, add allo_post config to router, which allow triggering the regular TCP decaping from a POST request payload.

Additional Description:  N/A

Risk Level: Low

Testing: Unit tests. Integration tests.

Docs Changes: Added

Release Notes: Added

Platform Specific Features: N/A

Fixes #14234 

cc @alyssawilk @mattklein123

### CODE REVIEW

**Comment by wlhee** [comment]
This PR is WIP.

Would appreciate feedback about API and naming conversion first.

File: api/envoy/extensions/filters/http/tcp_post/v3/tcp_post.proto [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import "udpa/annotations/status.proto";
+import "udpa/annotations/versioning.proto";
+
+option java_package = "io.envoyproxy.envoy.extensions.filters.http.tcp_post.v3";
+option java_outer_classname = "TcpPostProto";
+option java_multiple_files = true;
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: TCP Post]
+// Kill Request :ref:`configuration overview <config_http_filters_tcp_post>`.
```
Comment (lizan) [inline api/envoy/extensions/filters/http/tcp_post/v3/tcp_post.proto:16]: not `Kill Request`?

File: docs/root/configuration/http/http_filters/tcp_post_filter.rst [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+TCP streams over HTTP/2 POST requests. 
+
+Configuration
+-------------
+
+* This filter should be configured with the name *envoy.filters.http.tcp_post*.
+
+
+It converts POST requests back to CONNECT to trigger the regular TCP decapping. Therefore, it's
+typically used with another localhost listener that does the regular TCP decapping.
```
Comment (lizan) [inline docs/root/configuration/http/http_filters/tcp_post_filter.rst:16]: What's the reason that this filter only converts to CONNECT but not TCP directly?

File: api/envoy/extensions/filters/http/tcp_post/v3/tcp_post.proto (hunk 1 above)
Comment (wlhee) [reply api/envoy/extensions/filters/http/tcp_post/v3/tcp_post.proto]: Done.

File: docs/root/configuration/http/http_filters/tcp_post_filter.rst (hunk 2 above)
Comment (wlhee) [reply docs/root/configuration/http/http_filters/tcp_post_filter.rst:16]: The actual TCP decap is done by this regular envoy CONNECT support, not by this filter:

https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/http/upgrades#connect-support

Comment (lizan) [reply docs/root/configuration/http/http_filters/tcp_post_filter.rst:16]: I know TCP decap can be done by regular CONNECT support, but having an extra cluster/listener isn't really efficient because all traffic have to go through localhost twice. I'm asking what is the reason not do that within the filter (might including some refactor to reuse that part).

Comment (wlhee) [reply docs/root/configuration/http/http_filters/tcp_post_filter.rst:16]: Gotcha, @alyssawilk please chime in.

My thought is that I am not sure if this feature essential enough to change the core of regular TCP decamping, which is triggered by CONNECT.

File: api/envoy/extensions/filters/http/tcp_post/v3/tcp_post.proto [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+// [#extension: envoy.filters.http.tcp_post]
+
+// Configuration for TcpPost filter.
+message TcpPost {
+  // Specifies a set of request headers to trigger converting POST requests
+  // to CONNECT and set ``protocol: bytestream`` header.
+  //
+  // This is mainly used by the upstream proxy to coordinate with the downstream
+  // proxy that uses HTTP POST for TCP proxying.
+  repeated config.route.v3.HeaderMatcher headers = 1;
```
Comment (alyssawilk) [inline api/envoy/extensions/filters/http/tcp_post/v3/tcp_post.proto:26]: should we allow or reject changes to  : prefixed headers?
I'm inclined to say we should reject them, for consistency with other Envoy header modifiers unless you actually want to allow  host overrides, in which case we should explicitly comment and test.

File: source/common/tcp_proxy/upstream.cc [hunk 4]
```diff
@@ -244,31 +246,55 @@ void Http2Upstream::setRequestEncoder(Http::RequestEncoder& request_encoder, boo
...
+  });
+
+  if (!config_.use_post()) {
+    headers->addReference(Http::Headers::get().Protocol,
+                          Http::Headers::get().ProtocolValues.Bytestream);
+  }
+
+  for (const auto& header : config_.headers()) {
+    if (header.append().value()) {
+      headers->addCopy(Http::LowerCaseString(header.header().key()), header.header().value());
```
Comment (alyssawilk) [inline source/common/tcp_proxy/upstream.cc:263]: definitely being able to append to :-prefixed headers seems buggy

File: source/common/tcp_proxy/upstream.cc [hunk 5]
```diff
@@ -244,31 +246,55 @@ void Http2Upstream::setRequestEncoder(Http::RequestEncoder& request_encoder, boo
...
-    : HttpUpstream(callbacks, hostname) {}
+                             const TunnelingConfig& config)
+    : HttpUpstream(callbacks, config) {}
 
 void Http1Upstream::setRequestEncoder(Http::RequestEncoder& request_encoder, bool) {
   request_encoder_ = &request_encoder;
   request_encoder_->getStream().addCallbacks(*this);
 
   ASSERT(request_encoder_->http1StreamEncoderOptions() != absl::nullopt);
+  ASSERT(!config_.use_post());
```
Comment (alyssawilk) [inline source/common/tcp_proxy/upstream.cc:283]: do we have configuration validation and tests to ensure this config would be rejected?

File: source/common/tcp_proxy/upstream.cc [hunk 6]
```diff
@@ -244,31 +246,55 @@ void Http2Upstream::setRequestEncoder(Http::RequestEncoder& request_encoder, boo
...
       {Http::Headers::get().Method, "CONNECT"},
-      {Http::Headers::get().Host, hostname_},
+      {Http::Headers::get().Host, config_.hostname()},
   });
 
+  for (const auto& header : config_.headers()) {
+    if (header.append().value()) {
+      headers->addCopy(Http::LowerCaseString(header.header().key()), header.header().value());
+    } else {
+      headers->setCopy(Http::LowerCaseString(header.header().key()), header.header().value());
```
Comment (alyssawilk) [inline source/common/tcp_proxy/upstream.cc:294]: do we think it should be legal to configure host header here?  Again I'm inclined to say no, but if we allow we should call it out explicitly IMO.

File: docs/root/configuration/http/http_filters/tcp_post_filter.rst (hunk 2 above)
Comment (alyssawilk) [reply docs/root/configuration/http/http_filters/tcp_post_filter.rst]: I thought by adding a config option we were going to configure which methods trigger TCP decap.  CONNECT by default but others if configured.  WDYT?

File: api/envoy/extensions/filters/http/tcp_post/v3/tcp_post.proto (hunk 3 above)
Comment (mattklein123) [reply api/envoy/extensions/filters/http/tcp_post/v3/tcp_post.proto:26]: Can we try to intercept the generic matching support here? I think we have enough built that we should be ready to use it. @snowp should be your contact here and this will be a simple use case that we can make sure works end-to-end. Thank you!

File: api/envoy/extensions/filters/network/tcp_proxy/v3/tcp_proxy.proto [hunk 7]
```diff
@@ -68,6 +68,17 @@ message TcpProxy {
 
     // The hostname to send in the synthesized CONNECT headers to the upstream proxy.
     string hostname = 1 [(validate.rules).string = {min_len: 1}];
+
+    // Use POST method instead of CONNECT method to the tunnel the TCP stream.
```
Comment (mattklein123) [inline api/envoy/extensions/filters/network/tcp_proxy/v3/tcp_proxy.proto:72]:
```suggestion
    // Use POST method instead of CONNECT method to tunnel the TCP stream.
```

File: api/envoy/extensions/filters/network/tcp_proxy/v3/tcp_proxy.proto [hunk 8]
```diff
@@ -68,6 +68,17 @@ message TcpProxy {
...
+    // The 'protocol: bytestream' header is also NOT set to comply with the HTTP spec.
+    // Only HTTP/2 is supported.
+    //
+    // The upstream proxy is expected to convert POST back to CONNECT.
+    bool use_post = 2;
+
+    // Additional request headers to upstream proxy. This is mainly used to
+    // trigger upstream to convert POST requests back to CONNECT requests.
+    repeated config.core.v3.HeaderValueOption headers = 3;
```
Comment (mattklein123) [inline api/envoy/extensions/filters/network/tcp_proxy/v3/tcp_proxy.proto:81]:
```suggestion
    repeated config.core.v3.HeaderValueOption headers_to_add = 3;
```

To match other places we do this? Can you check the names of those fields?

File: docs/root/configuration/http/http_filters/tcp_post_filter.rst [hunk 9]
```diff
@@ -0,0 +1,16 @@
+.. _config_http_filters_tcp_post:
+
+TCP Post
+===============
+
+The TcpPost filter is used by the upstream proxy to coordinate with downstream proxy that proxies
+TCP streams over HTTP/2 POST requests.
```
Comment (mattklein123) [inline docs/root/configuration/http/http_filters/tcp_post_filter.rst:7]: Can you ref link to the TCP proxy field that enables this when used entirely within the context of Envoy?

File: docs/root/configuration/http/http_filters/tcp_post_filter.rst [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+The TcpPost filter is used by the upstream proxy to coordinate with downstream proxy that proxies
+TCP streams over HTTP/2 POST requests. 
+
+Configuration
+-------------
+
+* This filter should be configured with the name *envoy.filters.http.tcp_post*.
+
+It converts POST requests back to CONNECT to trigger the regular TCP decapping. Therefore, it's
+typically used with another localhost listener that handles the regular TCP decapping for CONNECT
```
Comment (mattklein123) [inline docs/root/configuration/http/http_filters/tcp_post_filter.rst:15]: I think a config example here would be worthwhile?

Also, stepping back, I thought at one point we discussed potentially building this directly into the CONNECT handling path in HCM so that we don't have to do a loop back? If we did that we don't need a filter and I think the entire thing is much simpler? Did we discard that idea for some reason?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks a few high level comments to get started!

/wait

File: docs/root/configuration/http/http_filters/tcp_post_filter.rst (hunk 2 above)
Comment (wlhee) [reply docs/root/configuration/http/http_filters/tcp_post_filter.rst]: SGTM. Let me give it a try.

Comment (wlhee) [reply docs/root/configuration/http/http_filters/tcp_post_filter.rst]: PTAL at the new API in api/envoy/config/route/v3/route_components.proto.

The binary change is also surprisingly simple :)

I also added example configs for encaping and terminating.

If this looks reasonable to you, I still move forward to add tests, docs and clean up the http filter.

Thanks!

cc @mattklein123

File: docs/root/configuration/http/http_filters/tcp_post_filter.rst (hunk 10 above)
Comment (wlhee) [reply docs/root/configuration/http/http_filters/tcp_post_filter.rst:15]: I added a new API to router, which could help avoid adding this http filter, please check the comment at the top, thx!

File: source/common/tcp_proxy/upstream.cc (hunk 4 above)
Comment (wlhee) [reply source/common/tcp_proxy/upstream.cc:263]: could you help point me to example to exclude : prefixed headers and host header?

File: source/common/tcp_proxy/upstream.cc (hunk 5 above)
Comment (wlhee) [reply source/common/tcp_proxy/upstream.cc:283]: that sounds good, I wonder how to do config validation when the tunnel config is in the listener but the http1 options is in the cluster?

https://github.com/envoyproxy/envoy/blob/842485709d651a6057b2ffb505ffced21173e004/configs/encapsulate_in_http1_connect.yaml#L28-L38

File: docs/root/configuration/http/http_filters/tcp_post_filter.rst (hunk 2 above)
Comment (mattklein123) [reply docs/root/configuration/http/http_filters/tcp_post_filter.rst]: I'm a strong +1 on just adding a config option to the route. I think it's much simpler and I think it's reasonable to have it be in the core. @lizan @alyssawilk?

/wait

Comment (alyssawilk) [reply docs/root/configuration/http/http_filters/tcp_post_filter.rst]: I think configuring the method in the upgrade configs is simpler and I'm fine with it being in core if you are :-)

Comment (wlhee) [reply docs/root/configuration/http/http_filters/tcp_post_filter.rst]: May I ask what is the "core" ?

IIUC, the API I added is in "route_components" - "upgrade_config" - "connect_connfig".

File: api/envoy/config/route/v3/route_components.proto [hunk 11]
```diff
@@ -745,6 +745,9 @@ message RouteAction {
     message ConnectConfig {
       // If present, the proxy protocol header will be prepended to the CONNECT payload sent upstream.
       core.v3.ProxyProtocolConfig proxy_protocol_config = 1;
+
+      // If set, the TCP proxy will expect POST payload as raw TCP.
```
Comment (lizan) [inline api/envoy/config/route/v3/route_components.proto:749]: comment is wrong. This is about the route.

Shall we name this `allow_post` and if it is true allow both POST and CONNECT be decapped?

File: api/envoy/config/route/v3/route_components.proto [hunk 12]
```diff
@@ -745,6 +745,9 @@ message RouteAction {
...
+      bool use_post = 2;
```
Comment (lizan) [inline api/envoy/config/route/v3/route_components.proto:750]: update the comment below for connect_config, because it is not only about `CONNECT` request.

File: docs/root/configuration/http/http_filters/tcp_post_filter.rst (hunk 2 above)
Comment (alyssawilk) [reply docs/root/configuration/http/http_filters/tcp_post_filter.rst]: oh sorry, actually in the default TCP/HTTP code which currently special cases CONNECT, rather than in an filter (which may or may not be included in a given Envoy build)

Comment (wlhee) [reply docs/root/configuration/http/http_filters/tcp_post_filter.rst]: I am still a bit confused. I wonder:

1) is route_components the right place to put the API? if not, what is the right place?
2) if yes, is connect_config the right place?
3) what does "special cases CONNECT" mean? I looked at the codes, it seems like router is the only place that chooses a tcp_stream for decaping. I don't see any places, such as http filter, http_connection_manager, would trigger the tcp decaping logic.

Comment (wlhee) [reply docs/root/configuration/http/http_filters/tcp_post_filter.rst]: Per offline discussion, the api added to router is preferred.

File: docs/root/configuration/http/http_filters/tcp_post_filter.rst (hunk 9 above)
Comment (wlhee) [reply docs/root/configuration/http/http_filters/tcp_post_filter.rst]: file removed.

**Comment by wlhee** [comment]
This PR is ready for a second round of review.

File: api/envoy/extensions/filters/http/tcp_post/v3/tcp_post.proto (hunk 3 above)
Comment (wlhee) [reply api/envoy/extensions/filters/http/tcp_post/v3/tcp_post.proto]: file removed.

File: source/common/tcp_proxy/upstream.cc (hunk 4 above)
Comment (wlhee) [reply source/common/tcp_proxy/upstream.cc:263]: It doesn't seem that other places in the code base that also add addition request headers have any check though ..

**Comment by mattklein123** [comment]
From an API perspective this LGTM at a high level. Can you merge main branch and we can take another look?

/wait

**Comment by wlhee** [comment]
Rebased.

File: docs/root/intro/arch_overview/http/upgrades.rst [hunk 13]
```diff
@@ -98,33 +98,45 @@ For an example of terminating connect, please see :repo:`configs/terminate_conne
...
-original TCP upstream, in this case to google.com.
+For HTTP/2 POST run `bazel-bin/source/exe/envoy-static --config-path configs/encapsulate_in_http2_post.yaml --base-id 1`
+and `bazel-bin/source/exe/envoy-static --config-path configs/terminate_http2_post.yaml`.
+
+In all cases you will be running a first Envoy listening for TCP traffic on port 10000 and
+encapsulating it in an HTTP CONNECT or HTTP POST request, and a second one listening on 10001,
+stripping the CONNECT headers (not needed for POST request), and forwarding the original TCP
+upstream, in this case to google.com.
 
-Envoy waits for the HTTP tunnel to be established (i.e. a successful response to the  CONNECT request is received),
```
Comment (alyssawilk) [inline docs/root/intro/arch_overview/http/upgrades.rst:129]: are a bunch of these edits the internal 80 char line limit? If so can you revert these?  :-)

File: docs/root/version_history/current.rst [hunk 14]
```diff
@@ -29,7 +29,46 @@ Removed Config or Runtime
 New Features
 ------------
 * access log: added the :ref:`formatters <envoy_v3_api_field_config.core.v3.SubstitutionFormatString.formatters>` extension point for custom formatters (command operators).
+* compression: the :ref:`compressor <envoy_v3_api_msg_extensions.filters.http.compressor.v3.Compressor>` filter adds support for compressing request payloads. Its configuration is unified with the :ref:`decompressor <envoy_v3_api_msg_extensions.filters.http.decompressor.v3.Decompressor>` filter with two new fields for different directions - :ref:`requests <envoy_v3_api_field_extensions.filters.http.compressor.v3.Compressor.request_direction_config>` and :ref:`responses <envoy_v3_api_field_extensions.filters.http.compressor.v3.Compressor.response_direction_config>`. The latter deprecates the old response-specific fields and, if used, roots the response-specific stats in `<stat_prefix>.compressor.<compressor_library.name>.<compressor_library_stat_prefix>.response.*` instead of `<stat_prefix>.compressor.<compressor_library.name>.<compressor_library_stat_prefix>.*`.
```
Comment (alyssawilk) [inline docs/root/version_history/current.rst:32]: looks like a bad merge?

File: source/common/tcp_proxy/upstream.cc (hunk 4 above)
Comment (alyssawilk) [reply source/common/tcp_proxy/upstream.cc:283]: https://github.com/envoyproxy/envoy/blob/master/source/common/router/header_parser.cc#L51

They do it on config ingress, not application, and I think we should do the same here.

File: source/common/tcp_proxy/upstream.cc (hunk 5 above)
Comment (alyssawilk) [reply source/common/tcp_proxy/upstream.cc:303]: At a high level I think it would be easier to allow this for HTTP/1 and HTTP/2 than try to restrict HTTP/1 but I'm fine with you doing either :-)

**Review by alyssawilk — COMMENTED** [review]
Yeah, I think the structure of this is much cleaner without the filter.  Thanks for that clean up!
Here's a few more thoughts, and we can go from there :-)

File: source/common/tcp_proxy/upstream.cc (hunk 5 above)
Comment (wlhee) [reply source/common/tcp_proxy/upstream.cc:303]: Done. I wonder does it violate HTTP/1 POST spec?

File: source/common/tcp_proxy/upstream.cc (hunk 4 above)
Comment (wlhee) [reply source/common/tcp_proxy/upstream.cc]: thx!

File: source/common/tcp_proxy/upstream.cc (hunk 6 above)
Comment (wlhee) [reply source/common/tcp_proxy/upstream.cc]: Added comment to the proto to call it out.

**Comment by wlhee** [comment]
Updated the PR to support HTTP/1.1 POST as suggested.

**Comment by alyssawilk** [comment]
Cool!  want to check out CI, and add the integration tests we talked about, and I'll take another look?

**Comment by wlhee** [comment]
Added integration tests, CI looks mostly okay.

File: api/envoy/extensions/filters/network/tcp_proxy/v3/tcp_proxy.proto [hunk 15]
```diff
@@ -68,6 +68,19 @@ message TcpProxy {
 
     // The hostname to send in the synthesized CONNECT headers to the upstream proxy.
     string hostname = 1 [(validate.rules).string = {min_len: 1}];
+
+    // Use POST method instead of CONNECT method to tunnel the TCP stream.
+    // The 'protocol: bytestream' header is also NOT set to comply with the HTTP spec.
+    // Only HTTP/2 is supported.
```
Comment (alyssawilk) [inline api/envoy/extensions/filters/network/tcp_proxy/v3/tcp_proxy.proto:74]: remove the limitation here?

File: docs/root/version_history/current.rst [hunk 16]
```diff
@@ -37,7 +37,10 @@ New Features
 * access log: added the :ref:`formatters <envoy_v3_api_field_config.core.v3.SubstitutionFormatString.formatters>` extension point for custom formatters (command operators).
 * http: added support for :ref:`:ref:`preconnecting <envoy_v3_api_msg_config.cluster.v3.Cluster.PreconnectPolicy>`. Preconnecting is off by default, but recommended for clusters serving latency-sensitive traffic, especially if using HTTP/1.1.
 * http: change frame flood and abuse checks to the upstream HTTP/2 codec to ON by default. It can be disabled by setting the `envoy.reloadable_features.upstream_http2_flood_checks` runtime key to false.
+* router: added :ref:`allow_post field <envoy_v3_api_field_config.route.v3.RouteAction.UpgradeConfig.ConnectConfig.allow_post>` for allowing POST payload as raw TCP.
 * tcp_proxy: add support for converting raw TCP streams into HTTP/1.1 CONNECT requests. See :ref:`upgrade documentation <tunneling-tcp-over-http>` for details.
+* tcp_proxy: added a :ref:`use_post field <envoy_v3_api_field_extensions.filters.network.tcp_proxy.v3.TcpProxy.TunnelingConfig.use_post>` for using HTTP/2 POST to proxy TCP streams.
```
Comment (alyssawilk) [inline docs/root/version_history/current.rst:42]: Also nix HTTP/2 here

File: api/envoy/extensions/filters/network/tcp_proxy/v3/tcp_proxy.proto [hunk 17]
```diff
@@ -68,6 +68,19 @@ message TcpProxy {
...
+    //
+    // The upstream proxy is expected to convert POST back to CONNECT.
+    bool use_post = 2;
+
+    // Additional request headers to upstream proxy. This is mainly used to
+    // trigger upstream to convert POST requests back to CONNECT requests.
+    // The hostname header can be overridden by this field.
```
Comment (alyssawilk) [inline api/envoy/extensions/filters/network/tcp_proxy/v3/tcp_proxy.proto:81]: @envoyproxy/api-shepherds thoughts on this?
Given we have the explicit hostname field I think we should disallow.  I'm inclined to say we disallow Host: overall given the  HTTP/1 and HTTP/2 weirdness.  WDYT of doing this globally, stating here that it's unsupported but unenforced and adding a TODO for me to enforce?

File: source/common/router/router.cc [hunk 18]
```diff
@@ -605,9 +605,19 @@ Filter::createConnPool(Upstream::ThreadLocalCluster& thread_local_cluster) {
...
-      route_entry_->connectConfig().has_value() &&
-      downstream_headers_->getMethodValue() == Http::Headers::get().MethodValues.Connect;
+
+  bool should_tcp_proxy = false;
+
+  if (route_entry_->connectConfig().has_value()) {
+    auto method = downstream_headers_->getMethodValue();
+    should_tcp_proxy = (method == Http::Headers::get().MethodValues.Connect);
+
+    // Allow using POST to proxy raw TCP is configured.
```
Comment (alyssawilk) [inline source/common/router/router.cc:615]: Maybe rephrase:
Allow POST for proxying raw TCP if it is configured.
?

File: api/envoy/extensions/filters/network/tcp_proxy/v3/tcp_proxy.proto (hunk 17 above)
Comment (alyssawilk) [inline api/envoy/extensions/filters/network/tcp_proxy/v3/tcp_proxy.proto:81]: Also I don't think we should allow overriding :method  - setting hostname but overriding with :authority here seems confusing.  Ditto use_post but setting :method.
I'd disallow, add config checks, and comment.

File: test/integration/tcp_tunneling_integration_test.cc [hunk 19]
```diff
@@ -400,6 +424,54 @@ TEST_P(TcpTunnelingIntegrationTest, Basic) {
...
+    filter->mutable_typed_config()->PackFrom(proxy_config);
+  });
+
+  initialize();
+
+  // Start a connection, and verify the upgrade headers are received upstream.
+  IntegrationTcpClientPtr tcp_client = makeTcpConnection(lookupPort("tcp_proxy"));
+  ASSERT_TRUE(fake_upstreams_[0]->waitForHttpConnection(*dispatcher_, fake_upstream_connection_));
+  ASSERT_TRUE(fake_upstream_connection_->waitForNewStream(*dispatcher_, upstream_request_));
+  ASSERT_TRUE(upstream_request_->waitForHeadersComplete());
```
Comment (alyssawilk) [inline test/integration/tcp_tunneling_integration_test.cc:448]: Check for POST here?

File: test/common/router/router_test.cc [hunk 20]
```diff
@@ -6253,6 +6253,28 @@ TEST_F(RouterTest, ConnectExplicitTcpUpstream) {
   router_.onDestroy();
 }
 
+TEST_F(RouterTest, PostExplicitTcpUpstream) {
```
Comment (alyssawilk) [inline test/common/router/router_test.cc:6256]: Maybe also test that if the config is not present, that newStream is called on the TCP pool?

**Review by alyssawilk — COMMENTED** [review]
Looking great! just a few more nits.

File: api/envoy/extensions/filters/network/tcp_proxy/v3/tcp_proxy.proto (hunk 17 above)
Comment (wlhee) [reply api/envoy/extensions/filters/network/tcp_proxy/v3/tcp_proxy.proto:81]: Sounds good, I will wait for https://github.com/envoyproxy/envoy/pull/14747 to merge in

File: test/common/router/router_test.cc (hunk 20 above)
Comment (wlhee) [reply test/common/router/router_test.cc:6256]: is this what ConnectExplicitTcpUpstream does?

File: api/envoy/extensions/filters/network/tcp_proxy/v3/tcp_proxy.proto (hunk 17 above)
Comment (htuch) [reply api/envoy/extensions/filters/network/tcp_proxy/v3/tcp_proxy.proto:80]: @alyssawilk yeah, I tend to agree. We already had one PR with `ext_proc` that we decided to disallow `host` override via header mutation.

Comment (htuch) [reply api/envoy/extensions/filters/network/tcp_proxy/v3/tcp_proxy.proto:80]: Tagging @gbrail and https://github.com/envoyproxy/envoy/pull/14385.

**Comment by alyssawilk** [comment]
host Pr landed so you should be good to merge that in :-)

File: source/common/tcp_proxy/upstream.cc (hunk 5 above)
Comment (wlhee) [reply source/common/tcp_proxy/upstream.cc]: Added HTTP/1 support and integration tests.

**Comment by wlhee** [comment]
Added config validation for headers_to_add, HTTP/1 support and updated integration_tests.

This PR is ready for review.

File: api/envoy/extensions/filters/network/tcp_proxy/v3/tcp_proxy.proto (hunk 17 above)
Comment (wlhee) [reply api/envoy/extensions/filters/network/tcp_proxy/v3/tcp_proxy.proto]: Thanks, added a config_test.

Comment (wlhee) [reply api/envoy/extensions/filters/network/tcp_proxy/v3/tcp_proxy.proto]: Should be disallowed now.

File: api/envoy/config/route/v3/route_components.proto [hunk 21]
```diff
@@ -741,10 +741,14 @@ message RouteAction {
         "envoy.api.v2.route.RouteAction.UpgradeConfig";
 
     // Configuration for sending data upstream as a raw data payload. This is used for
-    // CONNECT requests, when forwarding CONNECT payload as raw TCP.
+    // CONNECT requests, when forwarding CONNECT payload as raw TCP. It also allow using POST
```
Comment (alyssawilk) [inline api/envoy/config/route/v3/route_components.proto:744]: allow -> allows

File: test/common/router/router_test.cc (hunk 20 above)
Comment (alyssawilk) [reply test/common/router/router_test.cc:6332]: Sorry, I misspoke.  I meant that with connect config and without allow_post == true, that POST results in the http pool.

**Review by alyssawilk — COMMENTED** [review]
LGTM modulo that unit test
@lizan want to start on your pass?

File: test/common/router/router_test.cc [hunk 22]
```diff
@@ -6351,6 +6351,18 @@ TEST_F(RouterTest, PostExplicitTcpUpstream) {
   router_.onDestroy();
 }
 
+TEST_F(RouterTest, PostHttpUpstream) {
+  // Make sure POST request result in the HTTP pool.
```
Comment (alyssawilk) [inline test/common/router/router_test.cc:6355]: Sorry, can you add the same config above, without set_allow_post?
Basically I want this to verify your new "use connect" doesn't trigger with connect config present but allow_post false.

File: test/common/router/router_test.cc [hunk 23]
```diff
@@ -6329,6 +6329,40 @@ TEST_F(RouterTest, ConnectExplicitTcpUpstream) {
   router_.onDestroy();
 }
 
+TEST_F(RouterTest, PostExplicitTcpUpstream) {
+  // Explicitly configure an TCP upstream, to test factory creation.
+  cm_.thread_local_cluster_.cluster_.info_->upstream_config_ =
+      absl::make_optional<envoy::config::core::v3::TypedExtensionConfig>();
+  envoy::extensions::upstreams::http::tcp::v3::TcpConnectionPoolProto tcp_config;
```
Comment (alyssawilk) [inline test/common/router/router_test.cc:6336]: Ah, you should also be using the generic config here.

envoy::extensions::upstreams::http::tcp::v3::TcpConnectionPoolProto tcp_config;
->
envoy::extensions::upstreams::http::generic::v3::GenericConnectionPoolProto generic_config;

Comment (wlhee) [reply test/common/router/router_test.cc]: thanks, it works!

File: test/common/router/router_test.cc (hunk 22 above)
Comment (wlhee) [reply test/common/router/router_test.cc:6366]: done.

File: api/envoy/config/route/v3/route_components.proto [hunk 24]
```diff
@@ -741,10 +741,14 @@ message RouteAction {
         "envoy.api.v2.route.RouteAction.UpgradeConfig";
 
     // Configuration for sending data upstream as a raw data payload. This is used for
-    // CONNECT requests, when forwarding CONNECT payload as raw TCP.
+    // CONNECT requests, when forwarding CONNECT payload as raw TCP. It also allows using POST
+    // payload as raw TCP if configured.
```
Comment (lizan) [inline api/envoy/config/route/v3/route_components.proto:745]: This is not limited to `CONNECT` any more, would be better to re-word a bit.

File: api/envoy/extensions/filters/network/tcp_proxy/v3/tcp_proxy.proto [hunk 25]
```diff
@@ -68,6 +68,19 @@ message TcpProxy {
 
     // The hostname to send in the synthesized CONNECT headers to the upstream proxy.
     string hostname = 1 [(validate.rules).string = {min_len: 1}];
+
+    // Use POST method instead of CONNECT method to tunnel the TCP stream.
+    // The 'protocol: bytestream' header is also NOT set to comply with the HTTP spec.
+    //
+    // The upstream proxy is expected to convert POST back to CONNECT.
```
Comment (lizan) [inline api/envoy/extensions/filters/network/tcp_proxy/v3/tcp_proxy.proto:75]: convert POST back to raw TCP or CONNECT?

File: api/envoy/config/route/v3/route_components.proto (hunk 24 above)
Comment (wlhee) [reply api/envoy/config/route/v3/route_components.proto:745]: Reword, PTAL

**Comment by lizan** [comment]
Other than the docs nit LGTM, can you also resolve conflicts as well?

File: api/envoy/extensions/filters/network/tcp_proxy/v3/tcp_proxy.proto (hunk 25 above)
Comment (wlhee) [reply api/envoy/extensions/filters/network/tcp_proxy/v3/tcp_proxy.proto]: sorry, forgot to push.

**Review by lizan — DISMISSED** [review]

**Comment by lizan** [comment]
Can you fix coverage? it is below threshold.

**Comment by wlhee** [comment]
> Can you fix coverage? it is below threshold.

workkng on it.

**Comment by wlhee** [comment]
Hmm, asan failed to build  //test/common/router:router_test  withouut telling the reason. I wonder how to run the build it locally?

https://dev.azure.com/cncf/4684fb3d-0389-4e0b-8251-221942316e06/_apis/build/builds/64999/logs/165

**Comment by alyssawilk** [comment]
yeah sorry, CI is wedged: https://github.com/envoyproxy/envoy/pull/14844
Lizan, want to merge this regardless or wait for the fix / sync?

**Comment by wlhee** [comment]
I am also not sure why the "examples" CI also failed?
https://dev.azure.com/cncf/4684fb3d-0389-4e0b-8251-221942316e06/_apis/build/builds/64999/logs/200

**Comment by wlhee** [comment]
Run Mac CI seems broken ... not sure what is wrong:

https://dev.azure.com/cncf/4684fb3d-0389-4e0b-8251-221942316e06/_apis/build/builds/65182/logs/53

**Comment by wlhee** [comment]
CI seems very unstable recently. I just resolved the conflict of the version history, then multiple checks are failing :(

**Comment by wlhee** [comment]
CI is green now.

**Review by alyssawilk — APPROVED** [review]
@lizan can you re-stamp API and merge?

**Review by lizan — APPROVED** [review]

**Comment by wuhuizuo** [comment]
@wlhee any example codes(like C/go/rust)  for setting up the connecttions to remote envoy POST terminater?

**Comment by rtttech** [comment]
It's really awesome, how can I use this feature in istio?
